lib/cache/cache.go

```diff
@@ -1294,43 +1285,73 @@ func (c *Cache) GetNodes(ctx context.Context, namespace string, opts ...services
 	return rg.presence.GetNodes(ctx, namespace, opts...)
 }
 
+// getNodesWithFnCache implements TTL-based caching for the GetNodes endpoint.  All nodes that will be returned from the caching layer
+// must be cloned to avoid concurrent modification.
+func (c *Cache) getNodesWithFnCache(ctx context.Context, rg readGuard, namespace string, opts ...services.MarshalOption) ([]types.Server, error) {
+	ta := func(_ []types.Server) {} // compile-time type assertion
+	ni, err := c.fnCache.Get(ctx, getNodesCacheKey{namespace}, func() (interface{}, error) {
+		// use cache's close context instead of request context in order to ensure
+		// that we don't cache a context cancellation error.
+		nodes, err := rg.presence.GetNodes(c.ctx, namespace, opts...)
+		ta(nodes)
+		return nodes, err
+	})
+	if err != nil || ni == nil {
+		return nil, trace.Wrap(err)
+	}
+	cachedNodes := ni.([]types.Server)
+	ta(cachedNodes)
+	return cachedNodes, nil
+}
+
 // ListNodes is a part of auth.Cache implementation
 func (c *Cache) ListNodes(ctx context.Context, req proto.ListNodesRequest) ([]types.Server, string, error) {
-	// NOTE: we "fake" the ListNodes API here in order to take advantage of TTL-based caching of
-	// the GetNodes endpoint, since performing TTL-based caching on a paginated endpoint is nightmarish.
+	rg, err := c.read()
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if rg.IsCacheRead() {
+		// cache is healthy, delegate to the standard ListNodes implementation
+		return rg.presence.ListNodes(ctx, req)
+	}
+
+	// Cache is not healthy.  We need to take advantage of TTL-based caching.  Rather than caching individual page
+	// calls (very messy), we rely on caching the result of the GetNodes endpoint, and then "faking" pagination.
 
 	limit := int(req.Limit)
 	if limit <= 0 {
 		return nil, "", trace.BadParameter("nonpositive limit value")
 	}
 
-	nodes, err := c.GetNodes(ctx, req.Namespace)
+	cachedNodes, err := c.getNodesWithFnCache(ctx, rg, req.Namespace)
 	if err != nil {
 		return nil, "", trace.Wrap(err)
 	}
 
 	// trim nodes that precede start key
```

Comment:
Kinda... it would be recursively acquiring the readGuard lock, which is theoretically safe, but feels dirty.  Also, doing it this way allows us to only perform the deep copy on the portion of nodes we are actually returning.  The difference is likely negligible, but performance is a big concern for this method when you get into te 20-100k node range, so I'm taking every ms of gains I can get here.

Reply:
> Also, doing it this way allows us to only perform the deep copy on the portion of nodes we are actually returning

Shouldn't ListNodes do this too? Ie, only deep clone the page in question?

Up to you.

Reply:
That is what it does.  The below loop which builds the page only calls `DeepCopy` on the nodes that actually end up going into the page.